trade-finance-manager-api/src/v1/controllers/auth/auth-provider.js

```diff
@@ -0,0 +1,235 @@
+   * Prepares the auth code request parameters and initiates the first leg of auth code flow
+   * @param {String} csrfToken: CSRF token
+   * @param {Object} authCodeUrlRequestParams: Parameters for requesting an auth code url
+   * @param {Object} authCodeRequestParams: Parameters for requesting tokens using auth code
+   * @param {Object} msalInstance: MSAL instance
+   * @return {Object} PKCE codes, auth code request, login URL.
+   */
+  async getAuthCodeUrl({
+    csrfToken,
+    authCodeUrlRequestParams,
+    authCodeRequestParams,
+    msalInstance
+  }) {
+    try {
+      console.info('TFM auth service - getAuthCodeUrl');
+
+      /**
+       * Generate Proof Key of Code Exchange (PKCE) Codes,
+       * before starting the authorization flow.
+       */
+      const { verifier, challenge } = await this.cryptoProvider.generatePkceCodes();
+
+      /**
+       * Create a response object with:
+       * - CSRF token
+       * - Generated PKCE codes
+       * - Verifier
+       * - Challenge
+       */
+      let response = {
+        csrfToken,
+        pkceCodes: {
+          challengeMethod: 'S256',
+          verifier,
+          challenge,
+        },
+      };
+
+      /**
+       * By manipulating the request objects below before each request, we can obtain
+       * auth artifacts with desired claims. For more information, visit:
+       * https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#authorizationurlrequest
+       * https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#authorizationcoderequest
+       *
+       */
+      response = {
+        ...response,
+        authCodeUrlRequest: {
+          ...authCodeUrlRequestParams,
+          redirectUri: this.config.redirectUri,
+          responseMode: 'form_post', // recommended for confidential clients
+          codeChallenge: response.pkceCodes.challenge,
+          codeChallengeMethod: response.pkceCodes.challengeMethod,
+        },
+        authCodeRequest: {
+          ...authCodeRequestParams,
+          redirectUri: this.config.redirectUri,
+          code: '',
+        },
+      };
+
+      response.loginUrl = await msalInstance.getAuthCodeUrl(response.authCodeUrlRequest);
+
+      return response;
+    } catch (error) {
+      console.error('Error TFM auth service - getAuthCodeUrl %s', error);
+
+      throw new Error('Error TFM auth service - getAuthCodeUrl %s', error);
+    }
+  }
+
+  /**
+   * handleRedirect
+   * Modified Microsoft example, to meet out minimal needs.
+   * OpendId tokenResponse is enough for us (UKEF), it has all claims (aka fields) required for user data.
+   * We don't save Authorisation code because we don't need Access token.
+   * @param {Object} pkceCode: PKCE Code object
+   * @param {Object} origAuthCodeRequest: Original auth code request
+   * @param {String} code: authZ code
+   * @param {Object} req: Request object
+   * @returns {Object}
```

Comment:
this probably needs to be a promise of an object (side rant, this is why i dislike jsdocs, as this overrides the fact that this is a `async` function so we get errors when we try and `await handleRedirect`)

Reply:
JS docs are not very consistent, but improved them now for this PR.